### PR TITLE
executor: adjust the order of initialize plan, kvExecCounter and TopSQL

### LINKED ISSUE #51815: re-order the initialization of `StmtCtx` fields in `(*ExecStmt).Exec`
## Enhancement

In the `(a *ExecStmt) Exec`, some fields needed by the executors are not available after `buildExecutor`. Now they are shared by the same `StmtCtx` so it's fine to modify it after `buildExecutor`, but I think it's better to initialize them before `buildExecutor`.

### PR DESCRIPTION
**Pull request by YangKeao**
### What problem does this PR solve?

Issue Number: close #51815

This PR is part of #51618. I extract them because it's a logic change (but not simple refactor).

Problem Summary:

### What changed and how does it work?

I re-ordered the process of initialing plan digest and TopSQL related things (including KvExecCounter).

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [x] No need to test
  > - [ ] I checked and no code files have been changed.
  > They should be covered by existing tests.

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/executor/adapter.go [hunk 1]
```diff
@@ -1990,8 +1998,6 @@ func (a *ExecStmt) observeStmtBeginForTopSQL(ctx context.Context) context.Contex
 		// Always attach the SQL and plan info uses to catch the running SQL when Top SQL is enabled in execution.
 		if stats != nil {
 			stats.OnExecutionBegin(sqlDigestByte, planDigestByte)
-			// This is a special logic prepared for TiKV's SQLExecCount.
-			sc.KvExecCounter = stats.CreateKvExecCounter(sqlDigestByte, planDigestByte)
```
Comment (lcwangchao) [inline pkg/executor/adapter.go:1994]: why to delete this line?

Comment (YangKeao) [reply pkg/executor/adapter.go:1994]: It's not deleted. GitHub didn't fully update the PR Diff.  https://github.com/YangKeao/tidb/blob/adjust-stmtctx-fields-init-order/pkg/executor/adapter.go looks fine.

Comment (YangKeao) [reply pkg/executor/adapter.go:1994]: Change base backward and forward also cannot resolve this issue :facepalm:

**Review by xhebox — APPROVED** [review]

**Pull request merged**